parlai/tasks/dialogue_safety/base_agent.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from abc import ABC, abstractmethod
+import copy
+import os
+import json
+import random
+
+from parlai.core.message import Message
+from parlai.core.teachers import FixedDialogTeacher
+
+from .build import build, SINGLE_TURN_DATA
+
+
+OK_CLASS = '__ok__'
+NOT_OK_CLASS = '__notok__'
+
+
+class _BaseSafetyTeacher(FixedDialogTeacher, ABC):
+    """
+    Abstract parent class for single turn safety teachers.
+    Not meant to be a standalone teacher.
+    """
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        parser = parser.add_argument_group('Safety Teacher Args')
+        parser.add_argument(
+            '--round',
+            type=int,
+            default=1,
+            choices=[1, 2, 3],
+            help='Which round of data to use',
+        )
+        parser.add_argument(
+            '--round-only',
+            type='bool',
+            default=False,
+            help='if False, includes all rounds up to including the specified '
+            'round; if True, only includes data from the specified round',
+        )
+        return parser
+
+    def __init__(self, opt, shared=None):
+        build(opt['datapath'])  # download the data
+        self.opt = opt
+        self.data_path = os.path.join(
+            opt['datapath'], 'dialogue_safety', SINGLE_TURN_DATA
+        )
+
+        self.fixed_random = random.Random(42)
+        self.round = opt['round']
+        self.label_candidates = [NOT_OK_CLASS, OK_CLASS]
+
+        if shared and 'data' in shared:
+            self.data = shared['data']
+        else:
+            self._setup_data(opt['datatype'])
+
+        super().__init__(opt, shared)
+        self.reset()
+
+    @abstractmethod
+    def _load_data_dump(self):
```

Comment:
perhaps `raise NotImplementedError()` here as this is mandatory

Reply:
the decorator abstractmethod throws an error if a subclass is instantiated without this function implemented